src/osemgrep/reporting/Cli_json_output.ml

```diff
@@ -292,6 +292,61 @@ let cli_error_of_core_error (x : Out.core_error) : Out.cli_error =
 (*****************************************************************************)
 (* LATER: we should get rid of those intermediate Out.core_xxx *)
 
+(* This is a cursed function that calculates everything but the index part
+ * of the match_based_id. It's cursed because we need hashes to be exactly
+ * the same, but the algorithm used on the python side to generate
+ * the final string thats hashed has some python specific quirks.
+ *
+ * The way match based ID is calculated on the python side
+ * is as follows:
+ * (see https://github.com/returntocorp/semgrep/blob/2d34ce584d16c4e954349690a5f12fae877a94d6/cli/src/semgrep/rule.py#L289-L334)
+ * 1. Sort all top level keys (i.e pattern, patterns etc.) alphabetically
+ * 2. For each key: DFS the tree and find all pattern values (i.e. the rhs of pattern: <THING>)
+ * 3. Sort all pattern values alphabetically and concatenate them with a space
+ * 4. Concatenate all the sorted pattern values with a space
+ * 5. Hash the tuple `(sorted_pattern_values, path, rule_id)` w/ blake2b
+ * 6. Append the index of the match in the list of matches for the rule (see [index_match_based_ids])
+ *
+ * Austin: I wrote the initial match based ID, and this one below is a port of it.
+ * Looking back it seems like I ended up writing a roundabout version of the below algorithm
+ * which is how this function works
+ * 1. Same as before
+ * 2. for each rule: get all xpatterns, then sort them, and concatenate them with a space
+ * 3. Sort all of step 2 results alphabetically and concatenate them with a space
+ * 4. Same as step 5 and 6
+ *
+ * Assumptions:
+ * Somewhere it seems that all keys are already sorted alphabetically when the rule is parsed.
+ * I have not seen this code. I simply have faith that it is true and this will not change.
+ *
+ * I'm also hoping that [interpolate_metavariables] works similar to what we do on the python side
+ *
+ * I have not tested this code beyond checking a bunch of examples manually.
+ *
+ * There's some weird thing we do w/ join mode. I am hoping that this doesn't matter irl
```

Comment:
yep. We should not care too much about join mode; we will soon delete it as Emma has a replacement for it (steps mode).

Reply:
Cool that's what I figured